### PR TITLE
Postgres/DB2/Oracle: Fix comma join `LATERAL`.

### LINKED ISSUE #5516: Postgres/DuckDB/DB2/Oracle: Fix `LATERAL` parsing with `CROSS JOIN`
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

The `LATERAL` keyword fails to parse the bracketed part as either a `table_expression` or `select_statement` and parses it as a general `expression` like a normal function.
- Always for DB2 or Oracle
- When using comma cross joins in Postgres or DuckDB

### Expected Behaviour

`LATERAL` should parse as `keyword` instead of a `function` to allow the parsing of the `table_expression`.

### Observed Behaviour

```sh
echo 'select * from tbl1, lateral (values (1, 2, tbl1.id)) AS foo, lateral (select id, col1 from tbl2) as bar;' | sqlfluff parse --dialect postgres -

[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        select_statement:
[L:  1, P:  1]      |            select_clause:
[L:  1, P:  1]      |                keyword:                                      'select'
[L:  1, P:  7]      |                [META] indent:
[L:  1, P:  7]      |                whitespace:                                   ' '
[L:  1, P:  8]      |                select_clause_element:
[L:  1, P:  8]      |                    wildcard_expression:
[L:  1, P:  8]      |                        wildcard_identifier:
[L:  1, P:  8]      |                            star:                             '*'
[L:  1, P:  9]      |                [META] dedent:
[L:  1, P:  9]      |            whitespace:                                       ' '
[L:  1, P: 10]      |            from_clause:
[L:  1, P: 10]      |                keyword:                                      'from'
[L:  1, P: 14]      |                whitespace:                                   ' '
[L:  1, P: 15]      |                from_expression:
[L:  1, P: 15]      |                    [META] indent:
[L:  1, P: 15]      |                    from_expression_element:
[L:  1, P: 15]      |                        table_expression:
[L:  1, P: 15]      |                            table_reference:
[L:  1, P: 15]      |                                naked_identifier:             'tbl1'
[L:  1, P: 19]      |                    [META] dedent:
[L:  1, P: 19]      |                comma:                                        ','
[L:  1, P: 20]      |                whitespace:                                   ' '
[L:  1, P: 21]      |                from_expression:
[L:  1, P: 21]      |                    [META] indent:
[L:  1, P: 21]      |                    from_expression_element:
[L:  1, P: 21]      |                        table_expression:
[L:  1, P: 21]      |                            function:
[L:  1, P: 21]      |                                function_name:
[L:  1, P: 21]      |                                    function_name_identifier:  'lateral'
[L:  1, P: 28]      |                                whitespace:                   ' '
[L:  1, P: 29]      |                                bracketed:
[L:  1, P: 29]      |                                    start_bracket:            '('
[L:  1, P: 30]      |                                    [META] indent:
[L:  1, P: 30]      |                                    expression:
[L:  1, P: 30]      |                                        column_reference:
[L:  1, P: 30]      |                                            naked_identifier:  'values'
[L:  1, P: 36]      |                                    whitespace:               ' '
[L:  1, P: 37]      |                                    expression:
[L:  1, P: 37]      |                                        bracketed:
[L:  1, P: 37]      |                                            start_bracket:    '('
[L:  1, P: 38]      |                                            [META] indent:
[L:  1, P: 38]      |                                            numeric_literal:  '1'
[L:  1, P: 39]      |                                            comma:            ','
[L:  1, P: 40]      |                                            whitespace:       ' '
[L:  1, P: 41]      |                                            numeric_literal:  '2'
[L:  1, P: 42]      |                                            comma:            ','
[L:  1, P: 43]      |                                            whitespace:       ' '
[L:  1, P: 44]      |                                            column_reference:
[L:  1, P: 44]      |                                                naked_identifier:  'tbl1'
[L:  1, P: 48]      |                                                dot:          '.'
[L:  1, P: 49]      |                                                naked_identifier:  'id'
[L:  1, P: 51]      |                                            [META] dedent:
[L:  1, P: 51]      |                                            end_bracket:      ')'
[L:  1, P: 52]      |                                    [META] dedent:
[L:  1, P: 52]      |                                    end_bracket:              ')'
[L:  1, P: 53]      |                        whitespace:                           ' '
[L:  1, P: 54]      |                        alias_expression:
[L:  1, P: 54]      |                            keyword:                          'AS'
[L:  1, P: 56]      |                            whitespace:                       ' '
[L:  1, P: 57]      |                            naked_identifier:                 'foo'
[L:  1, P: 60]      |                    [META] dedent:
[L:  1, P: 60]      |                comma:                                        ','
[L:  1, P: 61]      |                whitespace:                                   ' '
[L:  1, P: 62]      |                from_expression:
[L:  1, P: 62]      |                    [META] indent:
[L:  1, P: 62]      |                    from_expression_element:
[L:  1, P: 62]      |                        table_expression:
[L:  1, P: 62]      |                            function:
[L:  1, P: 62]      |                                function_name:
[L:  1, P: 62]      |                                    function_name_identifier:  'lateral'
[L:  1, P: 69]      |                                whitespace:                   ' '
[L:  1, P: 70]      |                                bracketed:
[L:  1, P: 70]      |                                    start_bracket:            '('
[L:  1, P: 71]      |                                    [META] indent:
[L:  1, P: 71]      |                                    expression:
[L:  1, P: 71]      |                                        select_statement:
[L:  1, P: 71]      |                                            select_clause:
[L:  1, P: 71]      |                                                keyword:      'select'
[L:  1, P: 77]      |                                                [META] indent:
[L:  1, P: 77]      |                                                whitespace:   ' '
[L:  1, P: 78]      |                                                select_clause_element:
[L:  1, P: 78]      |                                                    column_reference:
[L:  1, P: 78]      |                                                        naked_identifier:  'id'
[L:  1, P: 80]      |                                                [META] dedent:
[L:  1, P: 80]      |                                    comma:                    ','
[L:  1, P: 81]      |                                    whitespace:               ' '
[L:  1, P: 82]      |                                    expression:
[L:  1, P: 82]      |                                        column_reference:
[L:  1, P: 82]      |                                            naked_identifier:  'col1'
[L:  1, P: 86]      |                                    whitespace:               ' '
[L:  1, P: 87]      |                                    unparsable:               !! Expected: 'Nothing here.'
[L:  1, P: 87]      |                                        word:                 'from'
[L:  1, P: 91]      |                                        whitespace:           ' '
[L:  1, P: 92]      |                                        word:                 'tbl2'
[L:  1, P: 96]      |                                    [META] dedent:
[L:  1, P: 96]      |                                    end_bracket:              ')'
[L:  1, P: 97]      |                        whitespace:                           ' '
[L:  1, P: 98]      |                        alias_expression:
[L:  1, P: 98]      |                            keyword:                          'as'
[L:  1, P:100]      |                            whitespace:                       ' '
[L:  1, P:101]      |                            naked_identifier:                 'bar'
[L:  1, P:104]      |                    [META] dedent:
[L:  1, P:104]      |    statement_terminator:                                     ';'
[L:  1, P:105]      |    newline:                                                  '\n'
[L:  2, P:  1]      |    [META] end_of_file:

==== parsing violations ====
L:   1 | P:  87 |  PRS | Line 1, Position 87: Found unparsable section: 'from tbl2'
```

```sh
echo 'select * from tbl1, lateral (values (1, 2, tbl1.id)) AS foo, lateral (select id, col1 from tbl2) as bar;' | sqlfluff parse --dialect db2 -

[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        select_statement:
[L:  1, P:  1]      |            select_clause:
[L:  1, P:  1]      |                keyword:                                      'select'
[L:  1, P:  7]      |                [META] indent:
[L:  1, P:  7]      |                whitespace:                                   ' '
[L:  1, P:  8]      |                select_clause_element:
[L:  1, P:  8]      |                    wildcard_expression:
[L:  1, P:  8]      |                        wildcard_identifier:
[L:  1, P:  8]      |                            star:                             '*'
[L:  1, P:  9]      |                [META] dedent:
[L:  1, P:  9]      |            whitespace:                                       ' '
[L:  1, P: 10]      |            from_clause:
[L:  1, P: 10]      |                keyword:                                      'from'
[L:  1, P: 14]      |                whitespace:                                   ' '
[L:  1, P: 15]      |                from_expression:
[L:  1, P: 15]      |                    [META] indent:
[L:  1, P: 15]      |                    from_expression_element:
[L:  1, P: 15]      |                        table_expression:
[L:  1, P: 15]      |                            table_reference:
[L:  1, P: 15]      |                                naked_identifier:             'tbl1'
[L:  1, P: 19]      |                    [META] dedent:
[L:  1, P: 19]      |                comma:                                        ','
[L:  1, P: 20]      |                whitespace:                                   ' '
[L:  1, P: 21]      |                from_expression:
[L:  1, P: 21]      |                    [META] indent:
[L:  1, P: 21]      |                    from_expression_element:
[L:  1, P: 21]      |                        table_expression:
[L:  1, P: 21]      |                            function:
[L:  1, P: 21]      |                                function_name:
[L:  1, P: 21]      |                                    function_name_identifier:  'lateral'
[L:  1, P: 28]      |                                whitespace:                   ' '
[L:  1, P: 29]      |                                bracketed:
[L:  1, P: 29]      |                                    start_bracket:            '('
[L:  1, P: 30]      |                                    [META] indent:
[L:  1, P: 30]      |                                    expression:
[L:  1, P: 30]      |                                        column_reference:
[L:  1, P: 30]      |                                            naked_identifier:  'values'
[L:  1, P: 36]      |                                    whitespace:               ' '
[L:  1, P: 37]      |                                    expression:
[L:  1, P: 37]      |                                        bracketed:
[L:  1, P: 37]      |                                            start_bracket:    '('
[L:  1, P: 38]      |                                            [META] indent:
[L:  1, P: 38]      |                                            numeric_literal:  '1'
[L:  1, P: 39]      |                                            comma:            ','
[L:  1, P: 40]      |                                            whitespace:       ' '
[L:  1, P: 41]      |                                            numeric_literal:  '2'
[L:  1, P: 42]      |                                            comma:            ','
[L:  1, P: 43]      |                                            whitespace:       ' '
[L:  1, P: 44]      |                                            column_reference:
[L:  1, P: 44]      |                                                naked_identifier:  'tbl1'
[L:  1, P: 48]      |                                                dot:          '.'
[L:  1, P: 49]      |                                                naked_identifier:  'id'
[L:  1, P: 51]      |                                            [META] dedent:
[L:  1, P: 51]      |                                            end_bracket:      ')'
[L:  1, P: 52]      |                                    [META] dedent:
[L:  1, P: 52]      |                                    end_bracket:              ')'
[L:  1, P: 53]      |                        whitespace:                           ' '
[L:  1, P: 54]      |                        alias_expression:
[L:  1, P: 54]      |                            [META] indent:
[L:  1, P: 54]      |                            keyword:                          'AS'
[L:  1, P: 56]      |                            whitespace:                       ' '
[L:  1, P: 57]      |                            naked_identifier:                 'foo'
[L:  1, P: 60]      |                            [META] dedent:
[L:  1, P: 60]      |                    [META] dedent:
[L:  1, P: 60]      |                comma:                                        ','
[L:  1, P: 61]      |                whitespace:                                   ' '
[L:  1, P: 62]      |                from_expression:
[L:  1, P: 62]      |                    [META] indent:
[L:  1, P: 62]      |                    from_expression_element:
[L:  1, P: 62]      |                        table_expression:
[L:  1, P: 62]      |                            function:
[L:  1, P: 62]      |                                function_name:
[L:  1, P: 62]      |                                    function_name_identifier:  'lateral'
[L:  1, P: 69]      |                                whitespace:                   ' '
[L:  1, P: 70]      |                                bracketed:
[L:  1, P: 70]      |                                    start_bracket:            '('
[L:  1, P: 71]      |                                    [META] indent:
[L:  1, P: 71]      |                                    expression:
[L:  1, P: 71]      |                                        select_statement:
[L:  1, P: 71]      |                                            select_clause:
[L:  1, P: 71]      |                                                keyword:      'select'
[L:  1, P: 77]      |                                                [META] indent:
[L:  1, P: 77]      |                                                whitespace:   ' '
[L:  1, P: 78]      |                                                select_clause_element:
[L:  1, P: 78]      |                                                    column_reference:
[L:  1, P: 78]      |                                                        naked_identifier:  'id'
[L:  1, P: 80]      |                                                [META] dedent:
[L:  1, P: 80]      |                                    comma:                    ','
[L:  1, P: 81]      |                                    whitespace:               ' '
[L:  1, P: 82]      |                                    expression:
[L:  1, P: 82]      |                                        column_reference:
[L:  1, P: 82]      |                                            naked_identifier:  'col1'
[L:  1, P: 86]      |                                    whitespace:               ' '
[L:  1, P: 87]      |                                    expression:
[L:  1, P: 87]      |                                        column_reference:
[L:  1, P: 87]      |                                            naked_identifier:  'from'
[L:  1, P: 91]      |                                    whitespace:               ' '
[L:  1, P: 92]      |                                    expression:
[L:  1, P: 92]      |                                        column_reference:
[L:  1, P: 92]      |                                            naked_identifier:  'tbl2'
[L:  1, P: 96]      |                                    [META] dedent:
[L:  1, P: 96]      |                                    end_bracket:              ')'
[L:  1, P: 97]      |                        whitespace:                           ' '
[L:  1, P: 98]      |                        alias_expression:
[L:  1, P: 98]      |                            [META] indent:
[L:  1, P: 98]      |                            keyword:                          'as'
[L:  1, P:100]      |                            whitespace:                       ' '
[L:  1, P:101]      |                            naked_identifier:                 'bar'
[L:  1, P:104]      |                            [META] dedent:
[L:  1, P:104]      |                    [META] dedent:
[L:  1, P:104]      |    statement_terminator:                                     ';'
[L:  1, P:105]      |    newline:                                                  '\n'
[L:  2, P:  1]      |    [META] end_of_file:

```

### How to reproduce

```sh
echo 'select * from tbl1, lateral (values (1, 2, tbl1.id)) AS foo, lateral (select id, col1 from tbl2) as bar;' | sqlfluff parse --dialect postgres -
echo 'select * from tbl1, lateral (values (1, 2, tbl1.id)) AS foo, lateral (select id, col1 from tbl2) as bar;' | sqlfluff parse --dialect duckdb -
echo 'select * from tbl1, lateral (values (1, 2, tbl1.id)) AS foo, lateral (select id, col1 from tbl2) as bar;' | sqlfluff parse --dialect db2 -
echo 'select * from tbl1, lateral (select id, col1 from tbl2);' | sqlfluff parse --dialect oracle -
```

### Dialect

DB2, Oracle, Postgres which also impacts DuckDB

### Version

Python 3.11.7
sqlfluff, version 3.0.0a4

### Configuration

defaults

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by keraion**
### Brief summary of the change made

fixes #5516 
- DB2/Oracle: Support `LATERAL` joins.
- Postgres/DuckDB: Support `LATERAL` joins after comma cross joins.
- Table expression alias excludes `JoinClauseSegment`.
- DB2/Postgres/DuckDB: Use `NATURAL JOIN` logic for condition-less `CROSS JOIN`.
- Updated a number of tests to valid, runnable queries.

### Are there any other side effects of this change that we should be aware of?
None

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_ansi.py [hunk 1]
```diff
@@ -539,17 +539,20 @@
...
-    NaturalJoinKeywordsGrammar=Sequence(
-        "NATURAL",
-        OneOf(
-            # Note that NATURAL joins do not support CROSS joins
-            "INNER",
-            Sequence(
-                OneOf("LEFT", "RIGHT", "FULL"),
-                Ref.keyword("OUTER", optional=True),
+    # Make OneOf to flip CROSS joins to natural style joins.
+    NaturalJoinKeywordsGrammar=OneOf(
```
[the remark this answers is not in the document]
Comment (keraion) [reply src/sqlfluff/dialects/dialect_ansi.py:543]: This was added to allow dialects such as postgres, duckdb, snowflake, and oracle which support both `NATURAL JOIN`s and `CROSS JOIN`s that don't allow for the `ON` or `USING` grammar. I'm open to suggestions if this isn't a recommended way.

Example added in this change in the postgres dialect:
```py
    NaturalJoinKeywordsGrammar=ansi_dialect.get_grammar(
        "NaturalJoinKeywordsGrammar"
    ).copy(
        insert=[Ref.keyword("CROSS")],
    ),
```

[the remark this answers is not in the document]
Comment (WittierDinosaur) [reply src/sqlfluff/dialects/dialect_ansi.py:543]: It feels like a little bit of a hack. I think we should have both NaturalJoinGrammar and CrossJoinGrammar as two separate things

[the remark this answers is not in the document]
Comment (keraion) [reply src/sqlfluff/dialects/dialect_ansi.py]: I've added a change to split out the join type grammar, but I'm not sure if I like it or not. On one hand it feels less hacky, on the other it seems potentially more difficult to understand. Once again, I'm open to suggestions on how best to proceed.

[the remark this answers is not in the document]
Comment (WittierDinosaur) [reply src/sqlfluff/dialects/dialect_ansi.py]: I like it! Thanks for doing the extra work. It is more verbose, but it's also less hacky

**Review by WittierDinosaur — APPROVED** [review]
LGTM

**Pull request merged**